### DISCUSSION TITLE
Trouble adding undo / redo

### DISCUSSION THREAD

**Discussion by deraleks1** [alexobviously/squares#13] [Q&A]
Heyo,

first off, thanks a lot publishing this flutter chess project.
I have been tinkering around a bit to develop a undo / redo button and 2-player mode (so without the AI) and started with the undo button but can't seem to figure out a previous state can be set. 
What I tried:
- Creating a List of BoardControllers inside game_page.dart
- Creating a list of GameControllers inside game_page.dart
- Just making a "reverse" move in makeMove inside game_controller.dart

Just wanted to ask how you would approach such task,

**Comment by alexobviously** [comment]
Hey @deraleks1
So the first thing - two player mode - I actually added to the example project just last week. It's actually the most recent commit on the `dev` branch: https://github.com/alexobviously/squares/commit/a7609b34afb1dfc5a33cd35787e782a662a82a5f. I'll be pushing this to `main` in the next version with a couple of other updates probably some time this week. Basically all you need to do for two players to play is not make a computer move, and update the legal moves on the board.

The other thing - an undo/redo system:
Undo is actually pretty easy if you're using Bishop. All you need to do is call `Game.undo()` and update the state from `Game.state`. Note though that this will fully undo your move though and you won't be able to redo, so if you do it this way, you'll have to store the move that you're undoing. Note that the `undo()` function actually does return you a move, so you can store that if you want.

Alternatively, you could leave `Game` alone, and build a manager of some sort to store `SquaresState` objects for each position the board has been in, and then navigate between them.

I am actually planning to build this at some point, either as part of the example or maybe even as part of the square bishop bridge package, so if you can't get this to work you can hold on a bit (maybe a couple of weeks?) and I'll have a solution you can use.

Also if you come up with anything cool related to this or anything else feel free to share!

**Reply by Mercutio1243** [reply]
Did you include the 2 player mode into main functionality or only into the example? Can't seem to find it in the main functionalities, I would therefore let myself be guided by the "complex example", on how to extend the functionality? 

Essentially, all I want to do is to enable the user to move both sides without the board rotating.

**Reply by Mercutio1243** [reply]
In case somebody is interested, I found the following workaround to enable the user to move both sides of the squares board without the board rotating:

1) Extend the bishop.Game with a new method, to create a modified squaresState that includes the information from which Players perspective the Board should be rotated, in addition to the information which player's turn in the game it is:

extension GameExtensions on bishop.Game {
  SquaresState squaresStateWithOrientation(int playerWithTurn, int playerWithBoardorientation) {
    return SquaresState(
      player: playerWithTurn,
      state: playState(playerWithTurn),
      size: size.toSquares(),
      board: boardState(playerWithBoardorientation),
      moves: squaresMoves(playerWithTurn),
      history: squaresHistory,
      hands: handSymbols,
      gates: gateSymbols,
    );
  }
}

2) Now use this modified method squaresStateWithOrientation() in your Squares BoardController, e.g., for onMove, replacing the regular squaresState() method, where you can not separately specify the boardOrientation, your squares widget should have:

  void onMove(Move move) async {
    int currentPlayer = state.player;
    int boardOrientationPlayer= 0; // adapt as needed
    bool result = globalDisplayedBishopChessGame.makeSquaresMove(move);
    if (result) {
      int newPlayer = currentPlayer == 0 ? 1 : 0;
      setState(() {
        state = globalDisplayedBishopChessGame.squaresStateWithOrientation(newPlayer, boardOrientationPlayer);
      });
    } else {
      print("Move not valid");
    }
  }
